Src/Base/AMReX_OpenMP.cpp

```diff
@@ -0,0 +1,175 @@
+#include <AMReX_OpenMP.H>
+#include <AMReX.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Print.H>
+
+#if defined(__APPLE__)
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#endif
+
+#if defined(_WIN32)
+#include <sysinfoapi.h>
+#endif
+
+#include <fstream>
+#include <iostream>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <vector>
+
+
+namespace amrex::OpenMP
+{
+    int
+    numUniquePhysicalCores ()
+    {
+        int ncores;
+
+#if defined(__APPLE__)
+        size_t len = sizeof(ncores);
+        // See hw.physicalcpu and hw.physicalcpu_max
+        //   https://developer.apple.com/documentation/kernel/1387446-sysctlbyname/determining_system_capabilities/
+        //   https://developer.apple.com/documentation/kernel/1387446-sysctlbyname
+        if (sysctlbyname("hw.physicalcpu", &ncores, &len, NULL, 0) == -1) {
+            if (system::verbose > 0) {
+                amrex::Print() << "numUniquePhysicalCores(): Error receiving hw.physicalcpu! "
+                               << "Defaulting to visible cores.\n";
+            }
+            ncores = int(std::thread::hardware_concurrency());
+        }
+#elif defined(__linux__)
+        std::set<std::vector<int>> uniqueThreadSets;
+        int cpuIndex = 0;
+
+        while (true) {
```

Comment:
Would it be simpler to grep (using `<regex>`) `/proc/cpuinfo`.
```
$ cat /proc/cpuinfo | grep "processor"
processor       : 0
processor       : 1
processor       : 2
processor       : 3
processor       : 4
processor       : 5
processor       : 6
processor       : 7

$ cat /proc/cpuinfo | grep "core"
core id         : 0
cpu cores       : 4
core id         : 1
cpu cores       : 4
core id         : 2
cpu cores       : 4
core id         : 3
cpu cores       : 4
core id         : 0
cpu cores       : 4
core id         : 1
cpu cores       : 4
core id         : 2
cpu cores       : 4
core id         : 3
cpu cores       : 4
```
I have 4 physical cored and 8 threads.

Reply:
`cpu cores` could be the right one.

Here is a
```
model name	: 12th Gen Intel(R) Core(TM) i9-12900H
```
```
$ cat /proc/cpuinfo | grep "processor"
processor	: 0
processor	: 1
processor	: 2
processor	: 3
processor	: 4
processor	: 5
processor	: 6
processor	: 7
processor	: 8
processor	: 9
processor	: 10
processor	: 11
processor	: 12
processor	: 13
processor	: 14
processor	: 15
processor	: 16
processor	: 17
processor	: 18
processor	: 19
axel@axel-dell:~$ cat /proc/cpuinfo | grep "core"
core id		: 0
cpu cores	: 14
core id		: 0
cpu cores	: 14
core id		: 4
cpu cores	: 14
core id		: 4
cpu cores	: 14
core id		: 8
cpu cores	: 14
core id		: 8
cpu cores	: 14
core id		: 12
cpu cores	: 14
core id		: 12
cpu cores	: 14
core id		: 16
cpu cores	: 14
core id		: 16
cpu cores	: 14
core id		: 20
cpu cores	: 14
core id		: 20
cpu cores	: 14
core id		: 24
cpu cores	: 14
core id		: 25
cpu cores	: 14
core id		: 26
cpu cores	: 14
core id		: 27
cpu cores	: 14
core id		: 28
cpu cores	: 14
core id		: 29
cpu cores	: 14
core id		: 30
cpu cores	: 14
core id		: 31
cpu cores	: 14
```

Effectively 14 physical cores. The first 12 logical cores use 2x SMT (6 physical). The next 8 are 1x SMT.

Reply:
Hm, does not work on RHEL (Lassen, POWER9, altivec supported, 4x SMT):
```
$ cat /proc/cpuinfo | grep "processor"
processor	: 0
processor	: 1
processor	: 2
processor	: 3
processor	: 4
processor	: 5
processor	: 6
processor	: 7
processor	: 8
processor	: 9
processor	: 10
processor	: 11
processor	: 12
processor	: 13
processor	: 14
processor	: 15
processor	: 16
processor	: 17
processor	: 18
processor	: 19
processor	: 20
processor	: 21
processor	: 22
processor	: 23
processor	: 24
processor	: 25
processor	: 26
processor	: 27
processor	: 28
processor	: 29
processor	: 30
processor	: 31
processor	: 32
processor	: 33
processor	: 34
processor	: 35
processor	: 36
processor	: 37
processor	: 38
processor	: 39
processor	: 40
processor	: 41
processor	: 42
processor	: 43
processor	: 44
processor	: 45
processor	: 46
processor	: 47
processor	: 48
processor	: 49
processor	: 50
processor	: 51
processor	: 52
processor	: 53
processor	: 54
processor	: 55
processor	: 56
processor	: 57
processor	: 58
processor	: 59
processor	: 60
processor	: 61
processor	: 62
processor	: 63
processor	: 64
processor	: 65
processor	: 66
processor	: 67
processor	: 68
processor	: 69
processor	: 70
processor	: 71
processor	: 72
processor	: 73
processor	: 74
processor	: 75
processor	: 76
processor	: 77
processor	: 78
processor	: 79
processor	: 80
processor	: 81
processor	: 82
processor	: 83
processor	: 84
processor	: 85
processor	: 86
processor	: 87
processor	: 88
processor	: 89
processor	: 90
processor	: 91
processor	: 92
processor	: 93
processor	: 94
processor	: 95
processor	: 96
processor	: 97
processor	: 98
processor	: 99
processor	: 100
processor	: 101
processor	: 102
processor	: 103
processor	: 104
processor	: 105
processor	: 106
processor	: 107
processor	: 108
processor	: 109
processor	: 110
processor	: 111
processor	: 112
processor	: 113
processor	: 114
processor	: 115
processor	: 116
processor	: 117
processor	: 118
processor	: 119
processor	: 120
processor	: 121
processor	: 122
processor	: 123
processor	: 124
processor	: 125
processor	: 126
processor	: 127
[huebl1@lassen708:~]$ cat /proc/cpuinfo | grep "core"
<empty>
```

Reply:
Same issue for the Power9 on Summit.

Reply:
The current implementation works with the Linux 4 and 5 kernels.

I am not sure what defines the `/proc/cpuinfo` output (OS/distribution or kernel), but it looks very diverse for the systems I tested and does not have the info we need in at least the Linux 4 kernel-based systems I tested.

Reply:
Didn't realize `/proc/cpuinfo` are so different on different machines.

Reply:
I had no idea either, today I learned...